File: LC083RemoveDuplicatesFromSortedList.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/remove-duplicates-from-sorted-list/
 */
public class LC083RemoveDuplicatesFromSortedList {
    public static class ListNode {
        int val;
        ListNode next;

        ListNode(int x) {
            val = x;
        }
    }

    public class Solution {
        public ListNode deleteDuplicates(ListNode head) {
            ListNode anchor = head;
            while (head != null) {
                while (head.next != null && head.val == head.next.val) {
                    head.next = head.next.next;
                }
                head = head.next;
            }
            return anchor;
        }
    }

    @Test
    public void test_1_1_2() throws Exception {
        ListNode n1a = new ListNode(1);
        ListNode n1b = new ListNode(1);
        ListNode n2 = new ListNode(2);
        n1a.next = n1b;
        n1b.next = n2;
        n2.next = null;
        ListNode result = new Solution().deleteDuplicates(n1a);
        assertEquals(1, result.val);
        assertEquals(2, result.next.val);
        assertEquals(null, result.next.next);
    }

    @Test
    public void test_1_1_2_3_3() throws Exception {
        ListNode n1a = new ListNode(1);
        ListNode n1b = new ListNode(1);
        ListNode n2 = new ListNode(2);
        ListNode n3a = new ListNode(3);
        ListNode n3b = new ListNode(3);
        n1a.next = n1b;
        n1b.next = n2;
        n2.next = n3a;
        n3a.next = n3b;
        n3b.next = null;
        ListNode result = new Solution().deleteDuplicates(n1a);
        assertEquals(1, result.val);
        assertEquals(2, result.next.val);
        assertEquals(3, result.next.next.val);
        assertEquals(null, result.next.next.next);
    }
}

File: algorithms/code/leetcode/lc083_remove_duplicates_from_sorted_list/lc083_remove_duplicates_from_sorted_list.py
# https://leetcode.com/problems/remove-duplicates-from-sorted-list/

from __future__ import annotations

import unittest


class ListNode:
    def __init__(
        self, val: int = 0, following: 'ListNode | None' = None
    ) -> None:
        self.val = val
        self.next = following


class Solution:
    @staticmethod
    def linked_to_listed(linked: ListNode | None) -> list:
        listed: list = []
        while linked is not None:
            listed.append(linked.val)
            linked = linked.next
        return listed

    @staticmethod
    def listed_to_linked(listed: list) -> ListNode | None:
        linked: ListNode | None = None
        for value in listed[::-1]:
            linked = ListNode(value, linked)
        return linked

    def deleteDuplicates(self, head: ListNode | None) -> ListNode | None:
        listed = Solution.linked_to_listed(head)
        deduplicated: list = []
        for value in listed:
            if not deduplicated or deduplicated[-1] != value:
                deduplicated.append(value)
        return Solution.listed_to_linked(deduplicated)


class TestCode(unittest.TestCase):
    def test_1_1_2(self) -> None:
        linked = Solution.listed_to_linked([1, 1, 2])
        result = Solution().deleteDuplicates(linked)
        self.assertListEqual([1, 2], Solution.linked_to_listed(result))

    def test_1_1_2_3_3(self) -> None:
        linked = Solution.listed_to_linked([1, 1, 2, 3, 3])
        result = Solution().deleteDuplicates(linked)
        self.assertListEqual([1, 2, 3], Solution.linked_to_listed(result))
